_posts/pl/2024-02-09-czy-wiesz-jak-bezprzerwowo-zmienic-schemat-w-mongodb.md

```diff
@@ -0,0 +1,81 @@
+
+Kiedy użyć?
+- kiedy nie może być przerwy w dostępie do systemu,
+- kiedy aktualizacja dokumentów może zająć godziny, dni lub tygodnie,
+- kiedy nie trzeba aktualizować wszystkich dokumentów do nowej wersji.
+
+## Przykład
+Pierwotnie zamodelowaliśmy dane kontaktowe w następujący sposób.
+
+```javascript
+{
+    "_id": "<ObjectId>",
+    "name": "Anakin Skywalker",
+    "home": "503-555-0000",
+    "work": "503-555-0010",
+    "mobile": "503-555-0120"
+}
+```
+
+Minęło trochę czasu, mamy już dużo dokumentów w naszej kolekcji, ale zmieniły się wymagania i musimy przechowywać więcej danych kontaktowych, związanych z nowymi formami komunikacji. Znamy na szczęście {% comment %}Zmienić na link jak już będzie znany[Attribute Pattern]({% post_url pl/2024-01-26-czy-wiesz-czym-jest-attribute-pattern-w-mongodb %}){% endcomment %}Attribute Pattern i decydujemy wprowadzić nie tylko nowe pola, ale i zastosować ten wzorzec - dodajemy pole `contact_method`.
+```javascript
+"contact_method": [
+    { "work": "503-555-0210" },
+    { "mobile": "503-555-0220" },
+    { "twitter": "@anakinskywalker" },
+    { "skype": "AlwaysWithYou" }
+]
+```
+
+Nie możemy sobie jednak pozwolić na przerwę w działaniu aplikacji i migrację od razu wszystkich dokumentów. Nowe dokumenty będą miały już tylko pole `contact_method`, a stare jeszcze dawne pola `home`, `work`, `mobile`.  Kod aplikacji będzie więc tworzył nowe dokumenty z polem `contact_method` przy dodawaniu. Może też ewentualnie aktualizować stare dokumenty przy okazji update'a.
+
+Wymaga to od nas dostosowania kodu aplikacji do dokumentów w dwóch różnych wersjach. Twórcy Mongo proponują bardzo prosty pomysł - dodanie dodatkowego pola `schema_version` w nowych dokumentach (lub zwiększenie wartości `schema_version`, jeśli już istnieje).
+
+```javascript
+{
+    "_id": "<ObjectId>",
+    "schema_version": "2",
+    "name": "Anakin Skywalker",
+    "contact_method": [
+        { "work": "503-555-0210" },
+        { "mobile": "503-555-0220" },
+        { "twitter": "@anakinskywalker" },
+        { "skype": "AlwaysWithYou" }
+    ]
+}
+```
+
+Mamy więc sytuację, gdy dokumenty są w dwóch różnych wersjach, a nasz kod musi radzić sobie z obsługą tych dwóch wersji i wskazane jest, żeby rozróżniać wersję właśnie po polu "schema_version" (w przeciwieństwie do wykrywania np. czy dane pole istnieje). Możemy równocześnie wykonać w tle masową aktualizację i zadecydować sami, kiedy ma się wydarzyć. W kolejnej wersji aplikacji możemy też zdecydować o pozbyciu się kodu obsługującego obydwie wersje.
```

Comment:
1. Mamy więc sytuację, gdy dokumenty są w dwóch różnych wersjach, a nasz kod musi radzić sobie z obsługą tych dwóch wersji i wskazane jest, żeby rozróżniać wersję właśnie po polu "schema_version" (w przeciwieństwie do wykrywania np. czy dane pole istnieje).

uprościłabym do:
Mamy więc dokumenty w dwóch różnych wersjach, a nasz kod musi poradzić sobie z obsługą obu wersji. Dodatkowo, wskazane jest, żeby rozróżniać wersje właśnie na podstawie obecności pola "schema_version".

przyznam szczerze, że tego co jest na końcu w nawiasie chyba nie rozumiem :(

Reply:
Podoba mi się to uproszczenie i je zastosuję, ale zmienia nieco sens tego zdania, więc wrzucę je w tej formie zastępując "obecności pola" -> "wartości pola".